src/webgpu/api/operation/memory_sync/buffer.spec.ts

```diff
@@ -0,0 +1,261 @@
+    const pipeline = this.createStorageWriteRenderPipeline(value);
+    const bindGroup = this.createBindGroup(pipeline, buffer);
+    const pass = this.beginSimpleRenderPass(encoder);
+    if (bundle) {
+      const bundleEncoder = this.device.createRenderBundleEncoder({
+        colorFormats: ['rgba8unorm'],
+      });
+      bundleEncoder.setBindGroup(0, bindGroup);
+      bundleEncoder.setPipeline(pipeline);
+      bundleEncoder.draw(1, 1, 0, 0);
+      const bundles = bundleEncoder.finish();
+      pass.executeBundles([bundles]);
+    } else {
+      pass.setPipeline(pipeline);
+      pass.setBindGroup(0, bindGroup);
+      pass.draw(1, 1, 0, 0);
+    }
+    pass.endPass();
+  }
+
+  // Write storage buffer via dispatch call in compute pass.
+  writeByComputePass(buffer: GPUBuffer, value: number, encoder: GPUCommandEncoder) {
+    const pipeline = this.createStorageWriteComputePipeline(value);
+    const bindGroup = this.createBindGroup(pipeline, buffer);
+    const pass = encoder.beginComputePass();
+    pass.setPipeline(pipeline);
+    pass.setBindGroup(0, bindGroup);
+    pass.dispatch(1);
+    pass.endPass();
+  }
+
+  // Issue write operation via render pass, compute pass, copy, etc.
+  issueWriteOp(
+    writeOp: string,
+    bundle: boolean,
+    buffer: GPUBuffer,
+    value: number,
+    encoder: GPUCommandEncoder
+  ) {
+    const isRenderPass = writeOp === 'render';
+    assert(!bundle || isRenderPass);
+    if (isRenderPass) {
+      this.writeByRenderPass(bundle, buffer, value, encoder);
+    } else {
+      assert(writeOp === 'compute');
+      this.writeByComputePass(buffer, value, encoder);
+    }
+  }
+
+  createCommandBufferAndIssueWriteOp(
+    writeOp: string,
+    bundle: boolean,
+    buffer: GPUBuffer,
+    value: number
+  ): GPUCommandBuffer {
+    const encoder = this.device.createCommandEncoder();
+    this.issueWriteOp(writeOp, bundle, buffer, value, encoder);
+    return encoder.finish();
+  }
+
+  verifyData(buffer: GPUBuffer, expectedValue: number) {
+    const bufferData = new Uint32Array(1);
+    bufferData[0] = expectedValue;
+    this.expectContents(buffer, bufferData);
+  }
+}
+
+export const g = makeTestGroup(BufferSyncTest);
+
+// Test write-after-write operations. The first write will write 1 into a storage buffer.
+// The second write will write 2 into the same storage buffer. So, expected data in buffer is 2.
+// The two writes can be in the same command buffer, or separate command buffers, or separate
+// queues. Each write operation can be done via render, compute, or copy, etc. If the write
+// operation is done by a render pass, it may use bundle.
+g.test('write_after_write,passes')
+  .params(
+    params()
+      // TODO (yunchao.he@intel.com): add multi-queue.
+      // Then the pbool 'sameCommandBuffer' can be a poption which consits of 3 options:
+      // same command buffer, separate command buffers, separate queues.
```

Comment:
Should probably test separate submit()s as well (this can be done now).

Reply:
True, actually I added tests for separate submits in my subsequent pr at my local side. I merged them into this one, including other write operations like copy, etc.(in addition to render and compute), and separate submits (in addition to one single cmdbuf, and separate cmdbufs).